middleware/middleware_test.go

```diff
@@ -0,0 +1,77 @@
+package middleware_test
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/middleware"
+	"github.com/stretchr/testify/require"
+)
+
+func TestMaxConcurrentRequestsMiddleware(t *testing.T) {
+
+	// create a handler to use as "next" which will verify the request
+	nextHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(time.Second * 1)
+		w.WriteHeader(http.StatusOK)
+	})
+
+	var tests = []struct {
+		name                  string
+		maxConcurrentRequests int
+		totalReq              int
+	}{
+		{
+			name:                  "max concurrent request: 2",
+			maxConcurrentRequests: 2,
+			totalReq:              10,
+		},
+		{
+			name:                  "max concurrent request: 1",
+			maxConcurrentRequests: 1,
+			totalReq:              10,
+		},
+		{
+			name:                  "max concurrent request: 10",
+			maxConcurrentRequests: 10,
+			totalReq:              10,
+		},
```

Comment:
[minor] and subjective.  I would prefer to see expected responses here, instead of computing them in:
```go
require.Equal(t, tt.totalReq-tt.maxConcurrentRequests, resp503, "actual 503 resp different than expected.")
```

Reply:
the reason I have written it like this is, to make it more intuitive that anything over allowed concurrentRequests would be 503. ` tt.totalReq-tt.maxConcurrentRequests` mathematically indicates that. @lvrach WDYT?